## 문제 11 유형(DataSet_11.csv 이용)

## 세계 각국의 행복지수를 비롯한 여러 정보를 조사한 DS리서치는 취합된 자료의 현황 파악 및 간단한 통계분석을 실시하고자 한다.

In [1]:
import pandas as pd

In [2]:
dataset11 = pd.read_csv('DataSet/DataSet_11.csv')

dataset11

Country  Happiness_Rank  Happiness_Score  year
0                 Switzerland               1            7.587  2015
1                     Iceland               2            7.561  2015
2                     Denmark               3            7.527  2015
3                      Norway               4            7.522  2015
4                      Canada               5            7.427  2015
..                        ...             ...              ...   ...
465                    Rwanda             151            3.471  2017
466                     Syria             152            3.462  2017
467                  Tanzania             153            3.349  2017
468                   Burundi             154            2.905  2017
469  Central African Republic             155            2.693  2017

[470 rows x 4 columns]

In [3]:
dataset11.columns

Index(['Country', 'Happiness_Rank', 'Happiness_Score', 'year'], dtype='object')

In [4]:
dataset11.dtypes

Country             object
Happiness_Rank       int64
Happiness_Score    float64
year                 int64
dtype: object

In [5]:
dataset11.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 470 entries, 0 to 469
Data columns (total 4 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Country          470 non-null    object 
 1   Happiness_Rank   470 non-null    int64  
 2   Happiness_Score  470 non-null    float64
 3   year             470 non-null    int64  
dtypes: float64(1), int64(2), object(1)
memory usage: 14.8+ KB


### 1. 분석을 위해 3년 연속 행복지수가 기록된 국가의 데이터를 사용하고자 한다. 3년 연속 데이터가 기록되지 않은 국가의 개수는?<br>- 국가명 표기가 한 글자라도 다른 경우 다른 국가로 처리하시오.<br>- 3년 연속 데이터가 기록되지 않은 국가 데이터는 제외하고 이를 향후 분석에서 활용하시오.<br>(답안 예시) 1

In [6]:
q1 = pd.pivot_table(dataset11, index=['Country'],
                    columns=['year'],
                    values='Happiness_Score')

q1

year          2015   2016   2017
Country                         
Afghanistan  3.575  3.360  3.794
Albania      4.959  4.655  4.644
Algeria      5.605  6.355  5.872
Angola       4.033  3.866  3.795
Argentina    6.574  6.650  6.599
...            ...    ...    ...
Venezuela    6.810  6.084  5.250
Vietnam      5.360  5.061  5.074
Yemen        4.077  3.724  3.593
Zambia       5.129  4.795  4.514
Zimbabwe     4.610  4.193  3.875

[166 rows x 3 columns]

In [7]:
q1['na_num'] = q1.isna().sum(axis=1)

q1

year          2015   2016   2017  na_num
Country                                 
Afghanistan  3.575  3.360  3.794       0
Albania      4.959  4.655  4.644       0
Algeria      5.605  6.355  5.872       0
Angola       4.033  3.866  3.795       0
Argentina    6.574  6.650  6.599       0
...            ...    ...    ...     ...
Venezuela    6.810  6.084  5.250       0
Vietnam      5.360  5.061  5.074       0
Yemen        4.077  3.724  3.593       0
Zambia       5.129  4.795  4.514       0
Zimbabwe     4.610  4.193  3.875       0

[166 rows x 4 columns]

In [8]:
q1_2 = q1[q1['na_num']<1]

q1_2

year          2015   2016   2017  na_num
Country                                 
Afghanistan  3.575  3.360  3.794       0
Albania      4.959  4.655  4.644       0
Algeria      5.605  6.355  5.872       0
Angola       4.033  3.866  3.795       0
Argentina    6.574  6.650  6.599       0
...            ...    ...    ...     ...
Venezuela    6.810  6.084  5.250       0
Vietnam      5.360  5.061  5.074       0
Yemen        4.077  3.724  3.593       0
Zambia       5.129  4.795  4.514       0
Zimbabwe     4.610  4.193  3.875       0

[146 rows x 4 columns]

In [9]:
(q1['na_num']>0).sum()

20

#### 정답 : ***20***

### 2. (1번 산출물을 활용하여) 2017년 행복지수와 2015년 행복지수를 활용하여 국가별 행복지수 증감률을 산출하고 행복지수 증감률이 가장 높은 3개 국가를 행복지수가 높은 순서대로 차례대로 기술하시오.<br>증감률 = (2017년행복지수−2015년행복지수)/2<br><br>- 연도는 년월(YEAR_MONTH) 변수로부터 추출하며, 연도별 매출금액합계는 1월부터 12월까지의 매출 총액을 의미한다.<br>(답안 예시) Korea, Japan, China

In [10]:
q2 = q1_2.drop(columns='na_num')

q2

year          2015   2016   2017
Country                         
Afghanistan  3.575  3.360  3.794
Albania      4.959  4.655  4.644
Algeria      5.605  6.355  5.872
Angola       4.033  3.866  3.795
Argentina    6.574  6.650  6.599
...            ...    ...    ...
Venezuela    6.810  6.084  5.250
Vietnam      5.360  5.061  5.074
Yemen        4.077  3.724  3.593
Zambia       5.129  4.795  4.514
Zimbabwe     4.610  4.193  3.875

[146 rows x 3 columns]

In [11]:
q2['ratio'] = (q2[2017] - q2[2015]) / 2

q2

year          2015   2016   2017   ratio
Country                                 
Afghanistan  3.575  3.360  3.794  0.1095
Albania      4.959  4.655  4.644 -0.1575
Algeria      5.605  6.355  5.872  0.1335
Angola       4.033  3.866  3.795 -0.1190
Argentina    6.574  6.650  6.599  0.0125
...            ...    ...    ...     ...
Venezuela    6.810  6.084  5.250 -0.7800
Vietnam      5.360  5.061  5.074 -0.1430
Yemen        4.077  3.724  3.593 -0.2420
Zambia       5.129  4.795  4.514 -0.3075
Zimbabwe     4.610  4.193  3.875 -0.3675

[146 rows x 4 columns]

In [12]:
q2['ratio'].nlargest(3).index

Index(['Latvia', 'Romania', 'Togo'], dtype='object', name='Country')

#### 정답 : ***Latvia, Romania, Togo***

### 3. (1번 산출물을 활용하여) 년도별 행복지수 평균이 유의미하게 차이가 나는지 알아보고자 한다. 이와 관련하여 적절한 검정을 사용하고 검정통계량을 기술하시오.<br>- 해당 검정의 검정통계량은 자유도가 2인 F 분포를 따른다.<br>- 검정통계량은 소수점 넷째 자리까지 기술한다.<br>(답안 예시) 0.1234

In [13]:
# 3개 이상의 표본 → ANOVA

In [14]:
from statsmodels.formula.api import ols
from statsmodels.stats.anova import anova_lm

In [15]:
q3 = dataset11[dataset11['Country'].isin(q1_2.index.values)]

q3

Country  Happiness_Rank  Happiness_Score  year
0    Switzerland               1            7.587  2015
1        Iceland               2            7.561  2015
2        Denmark               3            7.527  2015
3         Norway               4            7.522  2015
4         Canada               5            7.427  2015
..           ...             ...              ...   ...
464         Togo             150            3.495  2017
465       Rwanda             151            3.471  2017
466        Syria             152            3.462  2017
467     Tanzania             153            3.349  2017
468      Burundi             154            2.905  2017

[438 rows x 4 columns]

In [16]:
anova = ols('Happiness_Score~C(year)', q3).fit()   # C(수치형 변수) : 강제적으로 범주형으로 인식
# ols 사용 시 변수명에서 스페이스, -, .이 포함된 변수는 사용 불가

In [17]:
anova_lm(anova)

df      sum_sq   mean_sq         F    PR(>F)
C(year)     2.0    0.011198  0.005599  0.004277  0.995732
Residual  435.0  569.472307  1.309132       NaN       NaN

In [18]:
# C(year)  : 집단 간
# Residual : 집단 내
# PR(>F)   : p-value   →   0.9957 > 0.05이므로 귀무가설 기각 X

In [19]:
# 적어도 그룹 i는 평균이 같지 않다.

#### 정답 : ***0.0042***

In [20]:
##### ols를 쓰면 다음이 가능! #####
from statsmodels.stats.multicomp import pairwise_tukeyhsd   # 다중 비교(사후분석)

In [21]:
q3_out = pairwise_tukeyhsd(q3['Happiness_Score'], q3['year'])

print(q3_out)

Multiple Comparison of Means - Tukey HSD, FWER=0.05
group1 group2 meandiff p-adj  lower  upper  reject
--------------------------------------------------
  2015   2016  -0.0112   0.9 -0.3261 0.3038  False
  2015   2017   -0.001   0.9 -0.3159  0.314  False
  2016   2017   0.0102   0.9 -0.3048 0.3251  False
--------------------------------------------------


In [22]:
##### 이것도 가능! #####
from scipy.stats import f_oneway

In [23]:
f_oneway(q1_2[2015].values, q1_2[2016].values, q1_2[2017].values)

F_onewayResult(statistic=0.004276725037689305, pvalue=0.9957324489944479)

## 문제 12 유형(DataSet_12.csv 이용)

## 직장인의 독서 실태를 분석하기 위해서 수도권 거주자 5000명을 대상으로 간단한 인적 사항과 연간 독서량 정보를 취합하였다.

In [24]:
import pandas as pd

In [25]:
dataset12 = pd.read_csv('DataSet/DataSet_12.csv')

dataset12

Age Gender  Dependent_Count Education_Level  is_Married  \
0      52      M                2              석사           1   
1      46      F                4              박사           1   
2      53      F                0              석사           0   
3      55      M                2              학사           1   
4      51      F                3              석사           0   
...   ...    ...              ...             ...         ...   
4995   43      F                4              석사           0   
4996   41      F                3              학사           1   
4997   34      F                4              고졸           0   
4998   56      F                1              고졸           0   
4999   35      M                3              석사           1   

      Read_Book_per_Year Income_Range  
0                     45            D  
1                     26            X  
2                     48            A  
3                     45            D  
4                     46            A  
...                  ...          ...  
4995                  36            X  
4996                  15            B  
4997                  30            A  
4998                  48            A  
4999                  25            C  

[5000 rows x 7 columns]

In [26]:
dataset12.columns

Index(['Age', 'Gender', 'Dependent_Count', 'Education_Level', 'is_Married',
       'Read_Book_per_Year', 'Income_Range'],
      dtype='object')

In [27]:
dataset12.dtypes

Age                    int64
Gender                object
Dependent_Count        int64
Education_Level       object
is_Married             int64
Read_Book_per_Year     int64
Income_Range          object
dtype: object

In [28]:
dataset12.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 7 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   Age                 5000 non-null   int64 
 1   Gender              5000 non-null   object
 2   Dependent_Count     5000 non-null   int64 
 3   Education_Level     5000 non-null   object
 4   is_Married          5000 non-null   int64 
 5   Read_Book_per_Year  5000 non-null   int64 
 6   Income_Range        5000 non-null   object
dtypes: int64(4), object(3)
memory usage: 273.6+ KB


### 1. 수치형 변수를 대상으로 피어슨 상관분석을 실시하고 연간 독서량과 가장 상관관계가 강한 변수의 상관계수를 기술하시오.<br>- 상관계수는 반올림하여 소수점 셋째 자리까지 기술하시오.<br>(답안 예시) 0.123

### 2. 석사 이상(석사 및 박사) 여부에 따라서 연간 독서량 평균이 유의미하게 다른지 가설 검정을 활용하여 알아보고자 한다. 독립 2표본 t검정을 실시했을 때 유의 확률(pvalue)의 값을 기술하시오.- 등분산 가정 하에서 <br>검정을 실시한다.<br>- 유의 확률은 반올림하여 소수점 셋째 자리까지 기술한다.<br>(답안 예시) 0.123

### 3. 독서량과 다른 수치형 변수의 관계를 다중선형회귀분석을 활용하여 알아보고자 한다. 연간 독서량을 종속변수, 나머지 수치형 자료를 독립변수로 한다. 이렇게 생성한 선형회귀 모델을 기준으로 다른 독립변수가 고정이면서 나이만 다를 때, 40살은 30살보다 독서량이 얼마나 많은가?<br>- 학사 이상이면서 소득 구간 정보가 있는 데이터만 사용하여 분석을 실시하시오.<br>- 결과값은 반올림하여 정수로 표기하시오.<br>(답안 예시) 1

In [29]:
var_list = dataset12.columns[dataset12.dtypes!='object'].drop('Read_Book_per_Year')

In [30]:
form = 'Read_Book_per_Year~'+'+'.join(var_list)

ols1 = ols(form, data=dataset12).fit()

In [31]:
ols1.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:     Read_Book_per_Year   R-squared:                       0.635
Model:                            OLS   Adj. R-squared:                  0.635
Method:                 Least Squares   F-statistic:                     2901.
Date:                Tue, 24 Aug 2021   Prob (F-statistic):               0.00
Time:                        18:06:44   Log-Likelihood:                -15057.
No. Observations:                5000   AIC:                         3.012e+04
Df Residuals:                    4996   BIC:                         3.015e+04
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
===================================================================================
                      coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------
Intercept          -0.3231      0.445     -0.726      0.468      -1.195       0.549
Age                 0.7894      0.009     91.923      0.000       0.773       0.806
Dependent_Count    -0.1150      0.054     -2.116      0.034      -0.221      -0.008
is_Married         -0.0719      0.140     -0.514      0.608      -0.346       0.203
==============================================================================
Omnibus:                       72.985   Durbin-Watson:                   2.026
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               76.307
Skew:                          -0.288   Prob(JB):                     2.69e-17
Kurtosis:                       3.183   Cond. No.                         303.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [32]:
q3_ans = 0.7894 * 10   # 40 - 30 = 10

q3_ans

7.894

#### 정답 : ***8***

## 문제 13 유형(DataSet_13_train.csv / DataSet_test_13.csv 이용)

## 전국의 데이터 분석가 2000명을 대상으로 이직 관련 설문조사를 실시하였다. 설문 대상자의 특성 및 이직 의사와 관련 인자를 면밀히 살펴보기 위해 다양한 분석을 실시하고자 한다.

In [33]:
import pandas as pd

In [34]:
dataset_13_train = pd.read_csv('DataSet/DataSet_13_train.csv')
dataset_13_test  = pd.read_csv('DataSet/DataSet_13_test.csv')

dataset_13_train

city_development_index  gender  relevent_experience  \
0                      0.516    Male                    1   
1                      0.766    Male                    1   
2                      0.920    Male                    1   
3                      0.899    Male                    1   
4                      0.884    Male                    1   
...                      ...     ...                  ...   
1495                   0.920  Female                    1   
1496                   0.890  Female                    0   
1497                   0.920  Female                    0   
1498                   0.920  Female                    1   
1499                   0.802    Male                    1   

      enrolled_university education_level major_discipline  experience  \
0                       0        Graduate             STEM          13   
1                       0        Graduate             STEM          12   
2                       0         Masters             STEM          10   
3                       0        Graduate  Business Degree          16   
4                       0         Masters             STEM          20   
...                   ...             ...              ...         ...   
1495                    0         Masters             STEM           3   
1496                    0             Phd             STEM          20   
1497                    0         Masters             STEM           5   
1498                    0        Graduate             STEM          13   
1499                    0        Graduate             STEM           6   

      last_new_job  training_hours  target  
0                5              36       1  
1                5             264       0  
2                2              99       0  
3                1              20       0  
4                5              39       0  
...            ...             ...     ...  
1495             1             100       1  
1496             4              87       0  
1497             2              22       0  
1498             4              52       1  
1499             1              21       0  

[1500 rows x 10 columns]

In [35]:
dataset_13_train.columns

Index(['city_development_index', 'gender', 'relevent_experience',
       'enrolled_university', 'education_level', 'major_discipline',
       'experience', 'last_new_job', 'training_hours', 'target'],
      dtype='object')

In [36]:
dataset_13_train.dtypes

city_development_index    float64
gender                     object
relevent_experience         int64
enrolled_university         int64
education_level            object
major_discipline           object
experience                  int64
last_new_job                int64
training_hours              int64
target                      int64
dtype: object

In [37]:
dataset_13_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1500 entries, 0 to 1499
Data columns (total 10 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   city_development_index  1500 non-null   float64
 1   gender                  1500 non-null   object 
 2   relevent_experience     1500 non-null   int64  
 3   enrolled_university     1500 non-null   int64  
 4   education_level         1500 non-null   object 
 5   major_discipline        1500 non-null   object 
 6   experience              1500 non-null   int64  
 7   last_new_job            1500 non-null   int64  
 8   training_hours          1500 non-null   int64  
 9   target                  1500 non-null   int64  
dtypes: float64(1), int64(6), object(3)
memory usage: 117.3+ KB


### 1. (Dataset_13_train.csv를 활용하여) 경력과 최근 이직 시 공백기간의 상관관계를 보고자 한다. 남녀별 피어슨 상관계수를 각각 산출하고 더 높은 상관계수를 기술하시오.<br>- 상관계수는 반올림하여 소수점 둘째 자리까지 기술하시오.<br>(답안 예시) 0.12

In [38]:
Q1 = dataset_13_train.groupby(['gender'])[['experience', 'last_new_job']].corr().round(2)

Q1

experience  last_new_job
gender                                       
Female experience          1.00          0.45
       last_new_job        0.45          1.00
Male   experience          1.00          0.41
       last_new_job        0.41          1.00

#### 정답 : ***0.45***

### 2. (Dataset_13_train.csv를 활용하여) 기존 데이터 분석 관련 직무 경험과 이직 의사가 서로 관련이 있는지 알아보고자 한다. 이를 위해 독립성 검정을 실시하고 해당 검정의 p-value를 기술하시오.<br>- 검정은 STEM 전공자를 대상으로 한다.<br>- 검정은 충분히 발달된 도시(도시 개발 지수가 제 85 백분위수 초과)에 거주하는 사람을 대상으로 한다.<br>- 이직 의사 여부(target)은 문자열로 변경 후 사용한다.<br>- p-value는 반올림하여 소수점 둘째 자리까지 기술하시오.<br>(답안 예시) 0.12

In [39]:
from scipy.stats import chi2_contingency

In [40]:
Q2 = dataset_13_train.loc[(dataset_13_train['major_discipline']=='STEM') &
                          (dataset_13_train['city_development_index']>dataset_13_train['city_development_index'].quantile(0.85)), ]

Q2

city_development_index  gender  relevent_experience  \
6                      0.939  Female                    1   
14                     0.939    Male                    1   
20                     0.925    Male                    1   
38                     0.925    Male                    1   
56                     0.939    Male                    0   
...                      ...     ...                  ...   
1428                   0.926    Male                    0   
1448                   0.923  Female                    1   
1468                   0.925    Male                    0   
1469                   0.939  Female                    0   
1480                   0.923  Female                    1   

      enrolled_university education_level major_discipline  experience  \
6                       0        Graduate             STEM           9   
14                      0         Masters             STEM          20   
20                      0         Masters             STEM          16   
38                      0        Graduate             STEM          15   
56                      0        Graduate             STEM           3   
...                   ...             ...              ...         ...   
1428                    0         Masters             STEM          20   
1448                    0        Graduate             STEM          20   
1468                    1        Graduate             STEM           3   
1469                    1        Graduate             STEM           9   
1480                    0         Masters             STEM          12   

      last_new_job  training_hours  target  
6                5             148       0  
14               1              73       0  
20               2               8       0  
38               4             108       0  
56               2             109       0  
...            ...             ...     ...  
1428             5              15       0  
1448             2              56       1  
1468             2              13       0  
1469             0               9       1  
1480             1              81       0  

[145 rows x 10 columns]

In [41]:
# # 또는
# Q2 = dataset_13_train[(dataset_13_train['major_discipline']=='STEM') &
#                       (dataset_13_train['city_development_index']>dataset_13_train['city_development_index'].quantile(0.85))]
#                       # 맨 뒤에 , 사라짐

# Q2

In [42]:
stat, p, dof, exp_v = chi2_contingency(pd.crosstab(Q2['relevent_experience'], Q2['target']))

round(p, 2)

0.64

#### 정답 : ***0.64***

### 3. (Dataset_13_train.csv를 활용하여) 인사팀에서는 어떤 직원이 이직 의사를 가지고 있을지 사전에 파악하고 1:1 면담 등 집중 케어를 하고자 한다. 이를 위해 의사결정 나무를 활용하여 모델을 생성하고 그 정확도를 확인하시오.<br>- target을 종속변수로 하고 나머지 변수 중 String이 아닌 변수를 독립변수로 한다.<br>- 학습은 전부 기본값으로 실시한다.<br>- 평가는 "Dataset_13_test.csv" 데이터로 실시한다.<br>- 정확도는 반올림하여 소수점 둘째 자리까지 기술하시오.<br>(답안 예시) 0.12

In [43]:
from sklearn.tree import DecisionTreeClassifier

In [44]:
x_var = dataset_13_train.columns[dataset_13_train.dtypes!='object'].drop('target')

In [45]:
dt = DecisionTreeClassifier(random_state=123).fit(dataset_13_train[x_var], dataset_13_train['target'])

In [46]:
dt.score(dataset_13_test[x_var], dataset_13_test['target'])

0.672

#### 정답 : ***0.672***

## 문제 14 유형(DataSet_14.csv 이용)

## 온라인 교육업체 싱글캠퍼스에서 런칭한 교육 플랫폼을 보다 체계적으로 운영하기 위해 2014년부터 2016년 동안 개설된 강좌 2000개를 대상으로 강좌 실적 및 고객의 서비스 분석을 실시하려고 한다.

In [47]:
import pandas as pd

In [48]:
dataset14 = pd.read_csv('DataSet/DataSet_14.csv')

dataset14

id   published              subject               level  price  \
0        1  2015-07-10     Business Finance      Beginner Level    1.7   
1        2  2016-10-25     Business Finance          All Levels   17.4   
2        3  2015-06-16     Business Finance          All Levels    8.3   
3        4  2016-07-28     Business Finance          All Levels    7.0   
4        5  2016-05-18     Business Finance        Expert Level    8.3   
...    ...         ...                  ...                 ...    ...   
1995  1996  2016-07-02  Musical Instruments      Beginner Level    3.0   
1996  1997  2015-06-16     Business Finance      Beginner Level    1.7   
1997  1998  2016-09-21  Musical Instruments  Intermediate Level    5.2   
1998  1999  2016-12-31      Web Development      Beginner Level    8.3   
1999  2000  2015-10-28  Musical Instruments      Beginner Level    2.2   

      subscribers  reviews  lectures  duration  
0              56        8        18  2.000000  
1             336        6        13  0.683333  
2             871       79        59  2.000000  
3            1095       17        35  3.000000  
4             913      117        48  2.000000  
...           ...      ...       ...       ...  
1995           87       11        38  3.000000  
1996          244       13        21  2.500000  
1997            2        0        11  1.500000  
1998          502       62        59  5.000000  
1999            8        0        12  1.000000  

[2000 rows x 9 columns]

In [49]:
dataset14.columns

Index(['id', 'published', 'subject', 'level', 'price', 'subscribers',
       'reviews', 'lectures', 'duration'],
      dtype='object')

In [50]:
dataset14.dtypes

id               int64
published       object
subject         object
level           object
price          float64
subscribers      int64
reviews          int64
lectures         int64
duration       float64
dtype: object

In [51]:
dataset14.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2000 entries, 0 to 1999
Data columns (total 9 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   id           2000 non-null   int64  
 1   published    2000 non-null   object 
 2   subject      2000 non-null   object 
 3   level        2000 non-null   object 
 4   price        2000 non-null   float64
 5   subscribers  2000 non-null   int64  
 6   reviews      2000 non-null   int64  
 7   lectures     2000 non-null   int64  
 8   duration     2000 non-null   float64
dtypes: float64(2), int64(4), object(3)
memory usage: 140.8+ KB


### 1. 결제 금액이 1억 이상이면서 구독자의 리뷰 작성 비율이 10% 이상인 교육의 수는?<br>- 결제 금액은 강좌 가격에 구독자 수를 곱한 값이다.<br>- 리뷰 작성 비율은 리뷰 개수에 구독자 수를 나눈 값이다.<br>(답안 예시) 1

In [52]:
# 단위 고려

### 2. 강좌 가격이 비쌀수록 구독자 숫자는 줄어든다는 가설을 확인하기 위해 상관분석을 실시하고자 한다. 2016년 개설된 Web Development 강좌를 대상으로 강좌 가격과 구독자 수의 피어슨 상관관계를 기술하시오.<br>- 상관계수는 반올림하여 소수점 둘째 자리까지 기술하시오.<br>(답안 예시) 0.12

In [53]:
# 필터링 후 상관계수
# 날짜.dt.year

### 3. 유저가 서비스 사용에 익숙해지고 컨텐츠의 좋은 내용을 서로 공유하려는 경향이 전반적으로 증가하는 추세라고 한다. 이를 위해 먼저 강좌 개설 년도별 구독자의 리뷰 작성 비율의 평균이 강좌 개설 년도별로 차이가 있는지 일원 분산 분석을 통해서 알아보고자 한다. 이 때 검정통계량을 기술하시오.<br>- 검정통계량은 반올림하여 소수점 첫째 자리까지 기술하시오.<br>(답안 예시) 0.1

## 문제 15 유형(DataSet_05_Mart_POS.csv / DataSet_05_item_list.csv 이용)

## 원룸촌에 위치한 A마트는 데이터 분석을 통해 보다 체계적인 재고관리와 운영을 하고자 한다. 이를 위해 다음의 두 데이터 세트를 준비하였다.

In [54]:
import pandas as pd

In [55]:
pos = pd.read_csv('DataSet/DataSet_05_Mart_POS.csv')

pos

Member_number        Date itemDescription
0               1220  2015-01-01     canned beer
1               1220  2015-01-01       margarine
2               1220  2015-01-01       chocolate
3               1235  2015-01-01         sausage
4               1235  2015-01-01         sausage
...              ...         ...             ...
20483           4058  2015-12-30    cream cheese
20484           4565  2015-12-30     canned beer
20485           4565  2015-12-30     canned beer
20486           4863  2015-12-30         dessert
20487           4863  2015-12-30            curd

[20488 rows x 3 columns]

In [56]:
lst = pd.read_csv('DataSet/DataSet_05_item_list.csv')

lst

prod_id                prod_nm  alcohol  frozen
0          1         tropical fruit        0       0
1          2             whole milk        0       0
2          3              pip fruit        0       0
3          4       other vegetables        0       0
4          5             rolls/buns        0       0
..       ...                    ...      ...     ...
162      163         pudding powder        0       0
163      164            ready soups        0       0
164      165        make up remover        0       0
165      166         toilet cleaner        0       0
166      167  preservation products        0       0

[167 rows x 4 columns]

### 1. (DataSet_05_Mart_POS.csv를 활용하여) 가장 많은 제품이 팔린 날짜에 가장 많이 팔린 제품의 판매 개수는?<br>(답안 예시) 1

In [57]:
Q1_ans = pos['Date'].value_counts().nlargest(1)

Q1_ans

2015-01-21    96
Name: Date, dtype: int64

#### 정답 : ***96***

### 2. (DataSet_05_Mart_POS.csv, DataSet_05_item_list.csv를 활용하여) 고객이 주류 제품을 구매하는 요일이 다른 요일에 비해 금요일과 토요일이 많을 것이라는 가설을 세웠다. 이를 확인하기 위해 금요일과 토요일의 일별 주류제품 구매 제품 수 평균과 다른 요일의 일별 주류제품 구매 제품 수 평균이 서로 다른지 비교하기 위해 독립 2표본 t검정을 실시하시오. 해당 검정의 p-value를 기술하시오.<br>- 1분기(1월 ~ 3월) 데이터만 사용하여 분석을 실시하시오.<br>- 등분산 가정을 만족하지 않는다는 조건 하에 분석을 실시하시오.<br>- p-value는 반올림하여 소수점 둘째 자리까지 기술하시오.<br>(답안 예시) 0.12

In [58]:
# 1. 데이터셋 결함
# 2. 요일
# 3. 1분기 데이터 필터링(월 단위 호출)

In [59]:
q2 = pos.copy()

q2.dtypes

Member_number       int64
Date               object
itemDescription    object
dtype: object

In [60]:
pd.to_datetime(q2.Date).dt.year
pd.to_datetime(q2.Date).dt.month
pd.to_datetime(q2.Date).dt.day
pd.to_datetime(q2.Date).dt.day_name(locale='ko_kr')

0        목요일
1        목요일
2        목요일
3        목요일
4        목요일
        ... 
20483    수요일
20484    수요일
20485    수요일
20486    수요일
20487    수요일
Name: Date, Length: 20488, dtype: object

In [61]:
q2['month'] = pd.to_datetime(q2.Date).dt.month
q2['day']   = pd.to_datetime(q2.Date).dt.day_name(locale='ko_kr')

q2

Member_number        Date itemDescription  month  day
0               1220  2015-01-01     canned beer      1  목요일
1               1220  2015-01-01       margarine      1  목요일
2               1220  2015-01-01       chocolate      1  목요일
3               1235  2015-01-01         sausage      1  목요일
4               1235  2015-01-01         sausage      1  목요일
...              ...         ...             ...    ...  ...
20483           4058  2015-12-30    cream cheese     12  수요일
20484           4565  2015-12-30     canned beer     12  수요일
20485           4565  2015-12-30     canned beer     12  수요일
20486           4863  2015-12-30         dessert     12  수요일
20487           4863  2015-12-30            curd     12  수요일

[20488 rows x 5 columns]

In [62]:
q2_merge = pd.merge(q2, lst,
                    left_on='itemDescription',
                    right_on='prod_nm',
                    how='left')

q2_merge

Member_number        Date itemDescription  month  day  prod_id  \
0               1220  2015-01-01     canned beer      1  목요일       25   
1               1220  2015-01-01       margarine      1  목요일       70   
2               1220  2015-01-01       chocolate      1  목요일       14   
3               1235  2015-01-01         sausage      1  목요일       19   
4               1235  2015-01-01         sausage      1  목요일       19   
...              ...         ...             ...    ...  ...      ...   
20483           4058  2015-12-30    cream cheese     12  수요일       60   
20484           4565  2015-12-30     canned beer     12  수요일       25   
20485           4565  2015-12-30     canned beer     12  수요일       25   
20486           4863  2015-12-30         dessert     12  수요일       51   
20487           4863  2015-12-30            curd     12  수요일       45   

            prod_nm  alcohol  frozen  
0       canned beer        1       0  
1         margarine        0       0  
2         chocolate        0       0  
3           sausage        0       0  
4           sausage        0       0  
...             ...      ...     ...  
20483  cream cheese        0       0  
20484   canned beer        1       0  
20485   canned beer        1       0  
20486       dessert        0       0  
20487          curd        0       0  

[20488 rows x 9 columns]

In [63]:
q2_merge['week'] = 0
q2_merge.loc[q2_merge.day.isin(['금요일', '토요일']), 'week'] = 1   # 금, 토만 1

q2_merge

Member_number        Date itemDescription  month  day  prod_id  \
0               1220  2015-01-01     canned beer      1  목요일       25   
1               1220  2015-01-01       margarine      1  목요일       70   
2               1220  2015-01-01       chocolate      1  목요일       14   
3               1235  2015-01-01         sausage      1  목요일       19   
4               1235  2015-01-01         sausage      1  목요일       19   
...              ...         ...             ...    ...  ...      ...   
20483           4058  2015-12-30    cream cheese     12  수요일       60   
20484           4565  2015-12-30     canned beer     12  수요일       25   
20485           4565  2015-12-30     canned beer     12  수요일       25   
20486           4863  2015-12-30         dessert     12  수요일       51   
20487           4863  2015-12-30            curd     12  수요일       45   

            prod_nm  alcohol  frozen  week  
0       canned beer        1       0     0  
1         margarine        0       0     0  
2         chocolate        0       0     0  
3           sausage        0       0     0  
4           sausage        0       0     0  
...             ...      ...     ...   ...  
20483  cream cheese        0       0     0  
20484   canned beer        1       0     0  
20485   canned beer        1       0     0  
20486       dessert        0       0     0  
20487          curd        0       0     0  

[20488 rows x 10 columns]

In [64]:
q2_quater = q2_merge[q2_merge.month.isin([1, 2, 3])]

q2_quater

Member_number        Date  itemDescription  month  day  prod_id  \
0              1220  2015-01-01      canned beer      1  목요일       25   
1              1220  2015-01-01        margarine      1  목요일       70   
2              1220  2015-01-01        chocolate      1  목요일       14   
3              1235  2015-01-01          sausage      1  목요일       19   
4              1235  2015-01-01          sausage      1  목요일       19   
...             ...         ...              ...    ...  ...      ...   
5074           4678  2015-03-31   condensed milk      3  화요일       54   
5075           4678  2015-03-31       pot plants      3  화요일        6   
5076           4678  2015-03-31  cling film/bags      3  화요일       85   
5077           4678  2015-03-31             soda      3  화요일       38   
5078           4678  2015-03-31    shopping bags      3  화요일       71   

              prod_nm  alcohol  frozen  week  
0         canned beer        1       0     0  
1           margarine        0       0     0  
2           chocolate        0       0     0  
3             sausage        0       0     0  
4             sausage        0       0     0  
...               ...      ...     ...   ...  
5074   condensed milk        0       0     0  
5075       pot plants        0       0     0  
5076  cling film/bags        0       0     0  
5077             soda        0       0     0  
5078    shopping bags        0       0     0  

[5079 rows x 10 columns]

In [65]:
q2_tab = pd.pivot_table(q2_quater, index='Date',
                        columns='week',
                        values='alcohol',
                        aggfunc='sum')   # alcohol이 0과 1이기 때문에

q2_tab

week          0    1
Date                
2015-01-01  4.0  NaN
2015-01-02  NaN  3.0
2015-01-03  NaN  2.0
2015-01-04  4.0  NaN
2015-01-05  5.0  NaN
...         ...  ...
2015-03-27  NaN  5.0
2015-03-28  NaN  2.0
2015-03-29  5.0  NaN
2015-03-30  3.0  NaN
2015-03-31  4.0  NaN

[90 rows x 2 columns]

In [66]:
from scipy.stats import ttest_ind

In [67]:
ttest_ind(q2_tab[1].dropna(), q2_tab[0].dropna(), equal_var=True)   # True가 default → 등분산 가정

Ttest_indResult(statistic=-2.1325016442249423, pvalue=0.035750974474819806)

#### 정답 : ***0.04***

### 3. (DataSet_05_Mart_POS.csv를 활용하여) 1년 동안 가장 많이 판매된 10개 상품을 주력 상품으로 설정하고 특정 요일에 프로모션을 진행할지 말지 결정하고자 한다. 먼저 요일을 선정하기 전에 일원 분산 분석을 통하여 요일별 주력 상품의 판매 개수의 평균이 유의미하게 차이가 나는지 알아보고자 한다. 이와 관련하여 일원 분산 분석을 실시하고 p-value를 기술하시오.<br>- p-value는 반올림하여 소수점 둘째 자리까지 기술하시오.<br>(답안 예시) 0.12

In [68]:
pos.columns

Index(['Member_number', 'Date', 'itemDescription'], dtype='object')

In [69]:
top10 = pos['itemDescription'].value_counts().nlargest(10).index

top10

Index(['whole milk', 'other vegetables', 'rolls/buns', 'sausage', 'soda',
       'yogurt', 'tropical fruit', 'root vegetables', 'citrus fruit',
       'frankfurter'],
      dtype='object')

In [70]:
q3 = q2[q2['itemDescription'].isin(top10)]

q3

Member_number        Date   itemDescription  month  day
3               1235  2015-01-01           sausage      1  목요일
4               1235  2015-01-01           sausage      1  목요일
7               1422  2015-01-01    tropical fruit      1  목요일
18              2641  2015-01-01  other vegetables      1  목요일
19              2768  2015-01-01    tropical fruit      1  목요일
...              ...         ...               ...    ...  ...
20461           2997  2015-12-30    tropical fruit     12  수요일
20462           2997  2015-12-30       frankfurter     12  수요일
20466           2997  2015-12-30        whole milk     12  수요일
20473           3562  2015-12-30        rolls/buns     12  수요일
20474           3579  2015-12-30  other vegetables     12  수요일

[7866 rows x 5 columns]

In [71]:
q3_tab = pd.pivot_table(data=q3, index=['Date', 'day'],
                        values='itemDescription',
                        aggfunc='count').reset_index()

q3_tab

Date  day  itemDescription
0    2015-01-01  목요일               16
1    2015-01-02  금요일               13
2    2015-01-03  토요일               28
3    2015-01-04  일요일               21
4    2015-01-05  월요일               30
..          ...  ...              ...
359  2015-12-26  토요일               11
360  2015-12-27  일요일               31
361  2015-12-28  월요일               21
362  2015-12-29  화요일               27
363  2015-12-30  수요일               18

[364 rows x 3 columns]

In [72]:
from statsmodels.formula.api import ols
from statsmodels.stats.anova import anova_lm

In [73]:
ols1 = ols('itemDescription~day', q3_tab).fit()
q3_out = anova_lm(ols1)

q3_out

df        sum_sq    mean_sq        F    PR(>F)
day         6.0    219.527473  36.587912  0.86853  0.518128
Residual  357.0  15039.076923  42.126266      NaN       NaN

In [74]:
q3_out['PR(>F)'][0]

0.5181278037418005

#### 정답 : ***0.52***